Tasks/UsePythonVersionV0/Strings/resources.resjson/en-US/resources.resjson

```diff
@@ -16,7 +18,8 @@
   "loc.messages.PyPyNotFound": "Could not find an installation of PyPy%d in Agent.ToolsDirectory.",
   "loc.messages.ToolNotFoundMicrosoftHosted": "If this is a Microsoft-hosted agent, check that this image supports side-by-side versions of %s at %s.",
   "loc.messages.ToolNotFoundSelfHosted": "If this is a self-hosted agent, see how to configure side-by-side %s versions at %s.",
-  "loc.messages.VersionNotFound": "Version spec %s for architecture %s did not match any version in Agent.ToolsDirectory.",
-  "loc.messages.ExactVersionNotRecommended": "Specifying an exact version is not recommended on Microsoft-Hosted agents. Patch versions of Python can be replaced by new ones on Hosted agents without notice, which will cause builds to fail unexpectedly. It is recommended to specify only major or major and minor version (Example: `3` or `3.6`)",
-  "loc.messages.PythonVersionRetirement": "Python 3.5 has reached its end of life and will be removed from the ms-hosted images on January, 24. Please update Python version if you use ms-hosted agents. For more information about this - https://github.com/actions/virtual-environments"
+  "loc.messages.VersionNotFound": "Version spec %s for architecture %s did not match any version in Agent.ToolsDirectory or in python-versions registry (https://github.com/actions/python-versions).",
+  "loc.messages.PythonVersionRetirement": "Python 3.5 has reached its end of life and will be removed from the ms-hosted images on January, 24. Please update Python version if you use ms-hosted agents. For more information about this - https://github.com/actions/virtual-environments",
+  "loc.messages.DownloadNotFound": "Could not find Python matching spec %s in the python-versions registry. Beware that we don't support downloading python from registry on self-hosted agents.",
```

Comment:
Are there some reasons of this? Is this registry public?

Reply:
Actually as Maksim Zhukov stated, it should work on self-hosted agents, but they should be configured similarly to ms-hosted ones. I've clarified this a little bit in https://github.com/microsoft/azure-pipelines-tasks/commit/e4494dea0f8461e2120d0aaaa18c180d334e424d#diff-2ceaa2416857f8b99f47b61fed15299a46461677843e7907af3df3e071581f5fR89.

Reply:
@DaniilShmelev yes, as I see this registry is public - I don't see large obstacles to support it for self-hosted agent as well (that would be great to have in consistent for both hosted and self-hosted).
For 1st - we could probably add support for this in a separate ticket, for 2nd - as I see even for official [releases](https://www.python.org/downloads/) there are Linux/macOS/Windows distributions (there is a separate section with [Other](https://www.python.org/download/other/), but there are not so much for specific linux distributions), we could probably clarify which Linux packages supported/not supported for sure, and specify it in task docs/input description. 
Let's also try to figure out which exact issues with permissions do we see and decide if:
- Should we explicitly call chmod for some files?
- Are there some requirements we would need to specify in docs?

Reply:
I've looked into this and clarified the limitations of this feature on self-hosted agents:
On Ubuntu/Macos, there is a hardcoded path to the `libpython` library.
Ubuntu: `/opt/hostedtoolcache/Python/3.10.0/x64/lib/libpython3.10.so.1.0`
Macos: `/Users/runner/hostedtoolcache/Python/3.10.0/x64/lib/libpython3.10.dylib`
This means that on self-hosted agents, the tool directory will need to be the same as on ms-hosted ones.

On Windows, there is a different kind of problem - the installer prompts for Administrator permissions via UI. The only way for the feature to work on Windows is to run the agent with Administrator permissions.

In conclusion, it is possible to use this feature on self-hosted agents with severe limitations. We may add them to the documentation of this task.